---
title: 'Exploring relationship between HVGs and GC content in Jaitin et al. 2014 data'
author: "Stephanie Hicks"
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
        code_folding: hide
---

```{r load-libararies, message=FALSE, warning=FALSE}
library(scater)
library(scran)
library(cowplot)
library(dplyr)
library(gplots)
library(RColorBrewer)

baseDir <- "/net/irizarryfs01/srv/export/irizarryfs01_backed_up/share_root/shicks/projects/exploreHVGandGCbias"
```

# Load the Jaitin et al. data

Create an `SCESet` object. Uses UMIs, so we use counts per million (CPM) 
in `exprs` slot as transcript length is irrelevant. 

### Create `SCESet` object

Load data and create object
```{r make-sceset}
# library(devtools)
# install_github("jhsiao999/singleCellRNASeqMouseJaitinSpleen")
library(singleCellRNASeqMouseJaitinSpleen)
data(MouseJaitinSpleen)

# exclude ribosomal protein genes & that show significant batch specifc expression
# see http://compgenomics.weizmann.ac.il/tanay/?page_id=519 for more details
filter_genes <- c("M34473","abParts","M13680","Tmsb4x",
                    "S100a4","B2m","Atpase6","Rpl23","Rps18",
                    "Rpl13","Rps19","H2-Ab1","Rplp1","Rpl4",
                    "Rps26","EF437368") 
MouseJaitinSpleen <- MouseJaitinSpleen[!(rownames(fData(MouseJaitinSpleen)) %in% filter_genes), ]

# keep only cells with with 0 or more than 1 cell in the well
MouseJaitinSpleen <- MouseJaitinSpleen[, pData(MouseJaitinSpleen)$number_of_cells == 1]

table(pData(MouseJaitinSpleen)$group_name, pData(MouseJaitinSpleen)$plate_id)
table(pData(MouseJaitinSpleen)$sequencing_batch, pData(MouseJaitinSpleen)$plate_id)
table(pData(MouseJaitinSpleen)$sequencing_batch, pData(MouseJaitinSpleen)$amplification_batch)

# Create SCEset object
sce_jaitin_raw <- newSCESet(countData = exprs(MouseJaitinSpleen), phenoData = phenoData(MouseJaitinSpleen), 
                            featureData = featureData(MouseJaitinSpleen))
sce_jaitin_raw
```


### Calculate QC metrics
Define "feature controls" with ERCC spike-ins and mitochondrial genes. 
Calculate QC metrics.

```{r}
gene_names <- rownames(fData(MouseJaitinSpleen))
ercc_genes <- grep("ERCC", gene_names)
is.ercc <- grepl("ERCC", gene_names)

sce_jaitin_raw <- calculateQCMetrics(sce_jaitin_raw, 
                          feature_controls=list(ERCC = ercc_genes))
setSpike(sce_jaitin_raw) <- "ERCC" # specifies which feature control sets in SCESet are spike-ins
summary(isSpike(sce_jaitin_raw))
```

### Filter cells and genes

Filter genes. Use `calcIsExprs()` and `is_exprs()` to require a gene to have 
at least one count in at least 25 cells (want to remove genes without sufficient
counts). 

```{r}
is_exprs(sce_jaitin_raw) <- calcIsExprs(sce_jaitin_raw, lowerDetectionLimit = 0,
    exprs_values = "counts")           # expressed is defined non-zero in counts 
keep_gene <- rowSums(is_exprs(sce_jaitin_raw)) >= 3
fData(sce_jaitin_raw)$use <- keep_gene # use these genes in normalization and downstream analyses
```


Filter low-quality cells 
```{r}
libsize.drop <- isOutlier(sce_jaitin_raw$total_counts, nmads=3, type="lower", log=TRUE) # number molecules counted
feature.drop <- isOutlier(sce_jaitin_raw$total_features, nmads=3, type="lower", log=TRUE)  # number of features
spike.drop <- isOutlier(sce_jaitin_raw$pct_counts_feature_controls_ERCC, nmads=3, type="higher") # remove cells with high % ercc

sce_jaitin_raw$use <- !(libsize.drop | feature.drop | spike.drop | is.na(spike.drop))
```

Summary of filtering process 
```{r} 
data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop),
    BySpike=sum(spike.drop | is.na(spike.drop)), Remaining=sum(sce_jaitin_raw$use))
```



# Normalization

No mito genes to filter before identifying HVGs. 

```{r sizefactor-normalisation, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
## subset to form a QC'd version of the data

head(fData(sce_jaitin_raw))
sce_jaitin_qc <- sce_jaitin_raw[(fData(sce_jaitin_raw)$use | fData(sce_jaitin_raw)$is_feature_spike), 
                                sce_jaitin_raw$use] 
endog_genes <- !fData(sce_jaitin_qc)$is_feature_spike

## size factor normalisation with scran 
qclust <- quickCluster(sce_jaitin_qc)
sce_jaitin_qc <- computeSumFactors(sce_jaitin_qc, clusters = qclust)
summary(sce_jaitin_qc$size_factor)

# wider scatter plot than just along the line; reflects DE between cell types in Jaitin data
plot(sizeFactors(sce_jaitin_qc), sce_jaitin_qc$total_counts/1e3, log="xy",
    ylab="Library size (thousands)", xlab="Size factor")

# some cell specific scaling factors were less than 0. Remove these
sce_jaitin_qc <- sce_jaitin_qc[, sce_jaitin_qc$size_factor > 0] 


# compute scaling factors (total counts of ERCC spikes) separately for ERCC spikes
sce_jaitin_qc <- computeSpikeFactors(sce_jaitin_qc, type="ERCC", general.use=FALSE)

# normalize data
sce_jaitin_qc <- normalize(sce_jaitin_qc)

```

 

# identify highly variable genes

```{r}
var.fit <- trendVar(sce_jaitin_qc, use.spikes=TRUE) # must use use.spikes=TRUE (default) bc use.spikes=FALSE assumes no DE across cells; this not the case with Jaitin data
decomp <- decomposeVar(sce_jaitin_qc, var.fit)
decomp$SYMBOL <- rownames(featureData(sce_jaitin_qc))
top.hvgs <- order(decomp$bio, decreasing=TRUE)
head(decomp[top.hvgs,], 10)

# we see the variance of ERCC spike-ins are much smaller than the var of endog genes
# this reflects the heterogeneity in gene expression across cells of different types.
# Also means we cannot fit a trend to the endog var only (i.e. cannot use.spikes=FALSE) in trendVar)
pdf(file.path(baseDir, "figs/MeanVar_relationship_Jaitin.pdf"), width = 6, height = 6)
par(mfrow=c(1,1))
plot(decomp$mean, decomp$total, xlab="Mean log-expression", ylab="Variance")
points(decomp[top.hvgs[1:1000],]$mean, decomp[top.hvgs[1:1000],]$total, col = "green")
points(var.fit$mean, var.fit$var, col="red", pch=16)
o <- order(decomp$mean)
lines(decomp$mean[o], decomp$tech[o], col="red", lwd=2)
dev.off()
```


Calculate gc content for every gene in mouse genome 
```{r}
library(BSgenome.Mmusculus.UCSC.mm10) # load BSgenome package for mouse 
genome <- BSgenome.Mmusculus.UCSC.mm10
library(TxDb.Mmusculus.UCSC.mm10.ensGene) # load TxDb object for mouse using ensembl genes 
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene

# extract DNA string for only exons in each mouse gene
gene_seqs1 <- extractTranscriptSeqs(genome, exonsBy(txdb, by = "gene"))

# function to calculate gc content 
gcContent <- function(x){
  alf <- alphabetFrequency(x, as.prob=TRUE)
  sum(alf[c("G", "C")]) 
}

# calculate gc content for every transcript in mouse genome
genedbNames <- names(gene_seqs1)
gcVals <- numeric(length(genedbNames))             # pre-allocate...
minLen <- pmin(width(gene_seqs1), 300)
Sys.time()
for (ind in seq_along(genedbNames)) { 
  gcVals[ind] <- gcContent(subseq(gene_seqs1[[ind]]))
}
Sys.time()

dat <- data.frame(ENSEMBL = genedbNames, "gc" = gcVals)
head(dat)
```

Convert from gene symbol to ENSEMBL name

```{r}
library(org.Mm.eg.db)
anno <- select(org.Mm.eg.db, keys=as.character(decomp$SYMBOL), keytype="SYMBOL", column="ENSEMBL") 
anno$ENSEMBL[grep("ERCC", anno$SYMBOL)] <- anno$SYMBOL[grep("ERCC", anno$SYMBOL)]

dat.joined <- left_join(anno, dat,  by = "ENSEMBL")

# add gc content of ERCC spike-ins
data <- read.table("http://tools.thermofisher.com/downloads/ERCC_Controls_Annotation.txt", header=TRUE, stringsAsFactors=FALSE)
gcValsERCC <- numeric(length(data$ERCC_ID))             # pre-allocate...
for (ind in seq_along(data$ERCC_ID)) { 
  gcValsERCC[ind] <- gcContent(subseq(DNAString(data$Sequence[ind])))
}

dat.joined[dat.joined$SYMBOL %in% data$ERCC_ID , ]$gc <- 
  gcValsERCC[match(dat.joined$SYMBOL[grep("ERCC", dat.joined$SYMBOL)], data$ERCC_ID)]




x <- left_join(decomp, dat.joined, by = "SYMBOL")
x_ercc <- x[grep("ERCC", x$SYMBOL), ]
x <- na.omit(x)


pdf(file.path(baseDir, "figs/VarandGCcontent_relationship_Jaitin.pdf"), width = 16, height = 5)
par(mfrow=c(1,4))
plot(x$gc,x$total, xlab = "GC content", ylab = "total variance", ylim = c(0, 0.4))
fit.lo <- loess(x$total ~ x$gc)
points(x$gc, predict(fit.lo), col = 2) 
points(x_ercc$gc, x_ercc$total, col = 6, pch = 16)

plot(x$gc,x$tech, xlab = "GC content", ylab = "technical variance", ylim = c(0, 0.4))
fit.lo <- loess(x$tech ~ x$gc)
points(x$gc, predict(fit.lo), col = 2)
points(x_ercc$gc, x_ercc$tech, col = 6, pch = 16)

plot(x$gc,x$bio, xlab = "GC content", ylab = "biologial variance", ylim = c(-.02, 0.4))
fit.lo <- loess(x$bio ~ x$gc)
points(x$gc, predict(fit.lo), col = 2)
points(x_ercc$gc, x_ercc$bio, col = 6, pch = 16)

x$bio_new <- x$bio - predict(fit.lo.gc)
plot(x$gc, x$bio_new, ylim=c(-0.07, 0.4), xlab = "GC content", 
     ylab = "biological variance (GC corrected)")
fit.lo.new <- loess( x$bio_new ~ x$gc)
points(x$gc, predict(fit.lo.new), col = 2) 

dev.off()





x$bio_order <- seq(1:nrow(x))
x$bio_new_order <- seq(1:nrow(x))

x[order(x$bio, decreasing = TRUE),]$bio_order <- seq(1:nrow(x))
x[order(x$bio_new, decreasing = TRUE),]$bio_new_order <- seq(1:nrow(x))

plot(x$bio_order, x$bio_new_order)
abline(0,1, col =2)
plot(x$bio_order, x$bio_new_order, xlim =c(1, 1000), ylim = c(0, 1000))
abline(0, 1, col =2)

x.bio <- x[order(x$bio, decreasing = TRUE),]
x.bio_new <- x[order(x$bio_new, decreasing = TRUE),]

library(ggplot2)
p1 <- x %>% ggplot(aes(x=bio_order, y=bio_new_order, col = gc)) + geom_point() + 
  xlab("old order ranked by Biological Variance") + 
  ylab("new order ranked by Biological Variance (GC corrected)") 
p2 <- x.bio_new[1:100,] %>% ggplot(aes(x=bio_order, y=bio_new_order, col = gc)) + geom_point() + 
    xlab("old order ranked by Biological Variance") + 
  ylab("new order ranked by Biological Variance (GC corrected)") 

p.out <- plot_grid(p1,p2)

pdf(file.path(baseDir, "figs/BioVarOrder_Jaitin.pdf"), width = 10, height = 5)
print(p.out) 
dev.off()

head(x.bio, 10)
head(x.bio_new, 10)

dim(x.bio[which(x.bio$FDR <= 0.05 & x.bio$bio >= .5),])
dim(x.bio_new[which(x.bio_new$FDR <= 0.05 & x.bio_new$bio_new >= .5),])

g1 <- x.bio[which(x.bio$FDR <= 0.05 & x.bio$bio >= .5),]$SYMBOL
g2 <- x.bio_new[which(x.bio_new$FDR <= 0.05 & x.bio_new$bio_new >= .5),]$SYMBOL


setdiff(g1, g2)

gF <- as.factor(pData(sce_jaitin_qc)$group_name)
colors <- colorRampPalette(rev(brewer.pal(9, "PuOr")))(length(levels(gF)))
sidecols <- colors[ gF ]


pdf(file.path(baseDir, "figs/heatmap_Jaitin_celltype_g1.pdf"), width = 12, height = 8)
norm.exprs <- exprs(sce_jaitin_qc)[rownames(featureData(sce_jaitin_qc)) %in% g1, ] 
heat.vals <- norm.exprs - rowMeans(norm.exprs)
heatmap.2(heat.vals, col=bluered, symbreak=TRUE, 
          trace='none', cexRow=0.6, ColSideColors=sidecols)
legend("bottomleft", legend = levels(gF), col = colors, lty = 1, lwd = 10)
dev.off()

pdf(file.path(baseDir, "figs/heatmap_Jaitin_celltype_g2.pdf"), width = 12, height = 8)
norm.exprs <- exprs(sce_jaitin_qc)[rownames(featureData(sce_jaitin_qc)) %in% g2, ] 
heat.vals <- norm.exprs - rowMeans(norm.exprs)
heatmap.2(heat.vals, col=bluered, symbreak=TRUE, 
          trace='none', cexRow=0.6, ColSideColors=sidecols)
legend("left", legend = levels(gF), col = colors, lty = 1, lwd = 10)
dev.off()


```


